tests/31sync/09archived.pl

```diff
@@ -0,0 +1,193 @@
+test "Left rooms appear in the archived section of sync",
+   requires => [qw( first_api_client can_sync )],
+
+   check => sub {
+      my ( $http ) = @_;
+
+      my ( $user, $filter_id, $room_id );
+
+      matrix_register_user_with_filter( $http, {} )->then( sub {
+         ( $user, $filter_id ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_leave_room( $user, $room_id );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         my $room = $body->{rooms}{archived}{$room_id};
+         require_json_keys( $room, qw( event_map timeline state ));
+
+         Future->done(1);
+      });
+   };
+
+
+test "Newly left rooms appear in the archived section of incremental sync",
+   requires => [qw( first_api_client can_sync )],
+
+   check => sub {
+      my ( $http ) = @_;
+
+      my ( $user, $filter_id, $room_id, $next );
+
+     matrix_register_user_with_filter( $http, {} )->then( sub {
+         ( $user, $filter_id ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_sync( $user, filter => $filter_id );
+      })->then( sub {
+         my ($body ) = @_;
+
+         $next = $body->{next_batch};
+
+         matrix_leave_room( $user, $room_id );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id, since => $next );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         my $room = $body->{rooms}{archived}{$room_id};
+         require_json_keys( $room, qw( event_map timeline state ));
+
+         Future->done(1);
+      });
+   };
+
+
+test "Newly left rooms appear in the archived section of gapped sync",
+   requires => [qw( first_api_client can_sync )],
+
+   check => sub {
+      my ( $http ) = @_;
+
+      my ( $user, $filter_id, $room_id_1, $room_id_2, $next );
+
+      matrix_register_user_with_filter( $http, {} )->then( sub {
+         ( $user, $filter_id ) = @_;
+
+         Future->needs_all(
+            matrix_create_room( $user )->on_done( sub { ( $room_id_1 ) = @_; } ),
+            matrix_create_room( $user )->on_done( sub { ( $room_id_2 ) = @_; } ),
+         );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id );
+      })->then( sub {
+         my ($body ) = @_;
+
+         $next = $body->{next_batch};
+
+         matrix_leave_room( $user, $room_id_1 );
+      })->then( sub {
+         Future->needs_all( map {
+            matrix_send_room_message( $user, $room_id_2,
```

Comment:
It's unclear from the subsequent stages what these filler messages are doing. Also, you might want to pick a different word than "filler", due to its visual similarity and ease of accidental misreading/mistyping with "filter" :)